File: 프로그래머스/lv3/118669. 등산코스 정하기/등산코스 정하기.py
from heapq import heapify, heappush, heappop


def dijkstra(start, graph, summits_set, gates_set, intensity):
    q = []
    intensity[start] = 0
    heappush(q, (0, start))

    while q:
        dist, now = heappop(q)
        if now in summits_set or dist > intensity[now]:
            continue

        for c, v in graph[now]:
            if v in gates_set:
                continue

            cost = max(intensity[now], c)
            if intensity[v] <= cost:
                continue

            intensity[v] = cost
            heappush(q, (cost, v))


def solution(n, paths, gates, summits):
    _min, _max = 0, int(1e8)
    graph = [[] for _ in range(n + 1)]
    
    for a, b, c in paths:
        graph[a].append((c, b))
        graph[b].append((c, a))
    
    summits_set = set(summits)
    gates_set = set(gates)
    
    intensity = [int(1e8)] * (n + 1)
    for gate in gates:
        dijkstra(gate, graph, summits_set, gates_set, intensity)
    
    for s in summits_set:
        if intensity[s] < _max:
            _min, _max = s, intensity[s]
        elif intensity[s] == _max:
            _min = min(s, _min)
    
    return _min, _max

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.java
import java.util.*;

class Solution {

    static final int INF = (int) 1e8;

    private static int[] intensity;
    private static List<List<int[]>> graph;
    private static boolean[] isSummit;
    private static boolean[] isGate;

    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        graph = new ArrayList<>();
        intensity = new int[n + 1];
        isSummit = new boolean[n + 1];
        isGate = new boolean[n + 1];

        for (int i = 0; i <= n; i++) graph.add(new ArrayList<>());
        Arrays.fill(intensity, INF);

        for (int[] path : paths) {
            graph.get(path[0]).add(new int[]{path[2], path[1]});
            graph.get(path[1]).add(new int[]{path[2], path[0]});
        }

        for (int s : summits) isSummit[s] = true;
        for (int g : gates) {
            isGate[g] = true;
            intensity[g] = 0;
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        for (int g : gates) pq.offer(new int[]{0, g});

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int dist = cur[0];
            int now = cur[1];

            if (dist > intensity[now]) continue;
            if (isSummit[now]) continue;

            for (int[] next : graph.get(now)) {
                int c = next[0];
                int v = next[1];
                if (isGate[v]) continue;
                int nextCost = Math.max(intensity[now], c);
                if (nextCost < intensity[v]) {
                    intensity[v] = nextCost;
                    pq.offer(new int[]{nextCost, v});
                }
            }
        }

        Arrays.sort(summits);
        int minSummit = summits[0];
        int minIntensity = intensity[summits[0]];
        for (int i = 1; i < summits.length; i++) {
            if (intensity[summits[i]] < minIntensity) {
                minIntensity = intensity[summits[i]];
                minSummit = summits[i];
            }
        }

        return new int[]{minSummit, minIntensity};
    }
}